### PR TITLE
ring: add a randomized test

### PR DESCRIPTION
**Pull request by pav-kv**
Epic: none
Release note: none

### CODE REVIEW

File: pkg/util/container/ring/buffer_test.go [hunk 1]
```diff
@@ -131,3 +133,76 @@ func TestBuffer(t *testing.T) {
...
+	// Validate at each step that the reference queue matches the circular buffer.
+	var b Buffer[int]
+	check := func() {
+		require.Equal(t, b.Length(), len(queue))
+		for i, x := range queue {
+			require.Equal(t, x, b.At(i))
+		}
+	}
+
+	seed := int64(1234)
```
Comment (pav-kv) [inline pkg/util/container/ring/buffer_test.go:160]: Oups, need to make the seed random.

Comment (pav-kv) [reply pkg/util/container/ring/buffer_test.go]: Done.

**Comment by pav-kv** [comment]
```
$ dev test --stress --count=100 --filter=TestBufferRandom pkg/util/container/ring
...
//pkg/util/container/ring:ring_test                                      PASSED in 6.9s
  Stats over 100 runs: max = 6.9s, min = 5.1s, avg = 5.6s, dev = 0.4s
```

**Review by kvoli — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-:-OBkqkns73SrjwhEG2Lx:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/util/container/ring/buffer_test.go` line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBkq2u785Mf2btLNK9D:-OBkq2u785Mf2btLNK9E:bc3ahhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/679debe1842c33993df23177df2a9e52e66943d3/pkg/util/container/ring/buffer_test.go#L166)):*
> ```Go
> 
> 	operate := func(pushSkew float64, minSize, maxSize int, popEach int) {
> 		pops := 0    // number of planned items to pop
> ```

minor nit: I'd still use sentences above (and below) for these comments.
___
*[`pkg/util/container/ring/buffer_test.go` line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBkqXQxAhlFU6p5vroD:-OBkqXQxAhlFU6p5vroE:bnvc170) ([raw file](https://github.com/cockroachdb/cockroach/blob/679debe1842c33993df23177df2a9e52e66943d3/pkg/util/container/ring/buffer_test.go#L210)):*
> ```Go
> 	t.Run("pop-small", func(t *testing.T) { test(maxSize, 5) })
> 	t.Run("pop-med", func(t *testing.T) { test(maxSize, 25) })
> 	t.Run("pop-big", func(t *testing.T) { test(maxSize, 100) })
> ```

random nit: it might be interesting (and not that expensive) to add another subtest with larger primes for `popEach` and `maxSize`.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135298)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @kvoli and @sumeerbhola)

___
*[`pkg/util/container/ring/buffer_test.go` line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBkq2u785Mf2btLNK9D:-OBksnxhB9eXTqm9KVXT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/679debe1842c33993df23177df2a9e52e66943d3/pkg/util/container/ring/buffer_test.go#L166)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

minor nit: I'd still use sentences above (and below) for these comments.
</blockquote></details>

Done.
___
*[`pkg/util/container/ring/buffer_test.go` line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBkqXQxAhlFU6p5vroD:-OBksp-C9c0cvyu6zf5H:b-7txnlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/679debe1842c33993df23177df2a9e52e66943d3/pkg/util/container/ring/buffer_test.go#L210)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

random nit: it might be interesting (and not that expensive) to add another subtest with larger primes for `popEach` and `maxSize`.
</blockquote></details>

Added a bigger size + bigger pops. Not sure about it being prime, just set a random value where my fingers landed. Did you have a reason to make it prime?

File: pkg/util/container/ring/buffer_test.go [hunk 2]
```diff
@@ -131,3 +134,82 @@ func TestBuffer(t *testing.T) {
...
+		for i, x := range queue {
+			require.Equal(t, x, b.At(i))
+		}
+	}
+
+	seed := timeutil.Now().UnixNano()
+	t.Logf("Seed: %d", seed)
+	rng := rand.New(rand.NewSource(seed))
+
+	operate := func(pushSkew float64, minSize, maxSize int, popEach int) {
```
Comment (pav-kv) [inline pkg/util/container/ring/buffer_test.go:165]: Could have made it simpler:
- Set a fixed number of steps.
- A fixed prob of push/pop.
- And distribution for pop size.

Would need to make sure the thing does not blow up to a large size, by tuning the params.

LMK which sounds better, or if you have ideas how to stress `Buffer`-specific details more (e.g. the fact that it has a specific policy when to shrink).

**Comment by pav-kv** [comment]
bors r=kvoli

@sumeerbhola I'm merging for now, feel free to chime in with ideas [how to stress](https://github.com/cockroachdb/cockroach/pull/135298#discussion_r1844238474) specifics better.

**Review by sumeerbhola — COMMENTED** [review]
sorry about the delayed comments/questions.

> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-:-OBlSHktAM8OGqr5Qmp6:b-kzd7e4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/util/container/ring/buffer_test.go` line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlQ17LDeHz9YDFe5Xk:-OBlQ17LDeHz9YDFe5Xl:b2mblpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L173)):*
> ```Go
> 		schedule := func() {
> 			// mean == stddev == popEach
> 			nextPop = max(int(rng.NormFloat64()*float64(popEach)+float64(popEach)), 1)
> ```

I'm confused here. Isn't NormFloat64() potentially huge?
___
*[`pkg/util/container/ring/buffer_test.go` line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlQrLl7Jd9pTXF-MIh:-OBlQrLl7Jd9pTXF-MIi:b-6vkmdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L175)):*
> ```Go
> 			nextPop = max(int(rng.NormFloat64()*float64(popEach)+float64(popEach)), 1)
> 		}
> 		schedule()
> ```

I'm struggling to understand why the test is structured this way. Some code commentary would very much help the reader. Also, it's not obvious what `popEach` or `nextPop` mean, or what this is trying to `schedule`.
___
*[`pkg/util/container/ring/buffer_test.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlRIQW5eOiIuCyRHyR:-OBlRIQW5eOiIuCyRHyS:brd13gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L177)):*
> ```Go
> 		schedule()
> 
> 		for len(queue) > minSize && len(queue) < maxSize {
> ```

So as soon as one of the extremes is hit, it will return? And we could hit `minSize` even with a skew towards pushes, yes?
___
*[`pkg/util/container/ring/buffer_test.go` line 184 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlRWBr3p8QM8ZMiaPk:-OBlRWBr3p8QM8ZMiaPl:b6n2ejv) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L184)):*
> ```Go
> 				b.Push(x)
> 			} else if ln := len(queue) - pops; ln > 0 { // pop (if can)
> 				pops++
> ```

I don't understand what `pops` is. Code comment would really help. I would have expected that the `else if` case would actually do a pop. I suspect the interaction between `pops`, `nextPop` and `len(queue)` is somehow ensuring that we can exercise a pop that is larger than 1 entry, while respecting the `pushSkew` parameter. But I am struggling to put it together.
___
*[`pkg/util/container/ring/buffer_test.go` line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlS2YU9M_LQbJBuk96:-OBlS2YU9M_LQbJBuk97:b48w4no) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L193)):*
> ```Go
> 				schedule()
> 			}
> 			check()
> ```

btw, were you able to verify (by adding hacky instrumentation to Buffer) that shrinking gets exercised to a reasonable extent in runs of this test?
___
*[`pkg/util/container/ring/buffer_test.go` line 202 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlOip42YoA3lWJemC_:-OBlOip42YoA3lWJemCa:b-sro7ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L202)):*
> ```Go
> 		check()
> 		// Operate the buffer with skew towards pushes, until it reaches the max size.
> 		operate(0.8 /* pushSkew */, math.MinInt, maxSize, popEach)
> ```

nit: isn't this a negative `minSize` -- what does that mean?

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135298)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/util/container/ring/buffer_test.go` line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlQ17LDeHz9YDFe5Xk:-OBlTEl3-GbmAORAlK5Z:bew0ak2) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L173)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm confused here. Isn't NormFloat64() potentially huge?
</blockquote></details>

It returns from a normal distribution with mean=0, stddev=1, so it's unlikely going to be huge. And we scale it up to concentrate around `popEach`.

Could add a safe cap though.
___
*[`pkg/util/container/ring/buffer_test.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlRIQW5eOiIuCyRHyR:-OBlTlWWAIJ7NR8WeOtR:b-4p2hd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L177)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So as soon as one of the extremes is hit, it will return? And we could hit `minSize` even with a skew towards pushes, yes?
</blockquote></details>

Yes, stops on hitting the boundaries. But since we pass sentinel `MinInt` / `MaxInt` (see the callers), one of the boundaries is never hit. A hack, I suppose.

- `MinInt` is never hit because `len >= 0`. So the stopping condition is hitting the `maxSize`.
- `MaxInt` is never hit because the skew is downward. So the stopping condition is `minSize`.
___
*[`pkg/util/container/ring/buffer_test.go` line 184 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlRWBr3p8QM8ZMiaPk:-OBlUeqmAaPSvL-zAgse:b-r3d9fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L184)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand what `pops` is. Code comment would really help. I would have expected that the `else if` case would actually do a pop. I suspect the interaction between `pops`, `nextPop` and `len(queue)` is somehow ensuring that we can exercise a pop that is larger than 1 entry, while respecting the `pushSkew` parameter. But I am struggling to put it together.
</blockquote></details>

I wanted the "rate" of pushes to be `N * pushSkew` entries, and "rate" of pops be `N * (1 - pushSkew)` entries.

But because it's not interesting to do only one-entry pops, they are instead "planned"/"inteneded" to be popped when the next batched pop happens. So the "schedule" thing models when this batched pop happens, without affecting the overall push/pop rate.
___
*[`pkg/util/container/ring/buffer_test.go` line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlS2YU9M_LQbJBuk96:-OBlUN8V1ilDQRubmTan:b23fnbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L193)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

btw, were you able to verify (by adding hacky instrumentation to Buffer) that shrinking gets exercised to a reasonable extent in runs of this test?
</blockquote></details>

No, but would be good to do.
___
*[`pkg/util/container/ring/buffer_test.go` line 202 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/135298#-OBlOip42YoA3lWJemC_:-OBlUQ1F0yGHEfOet35s:bzhz051) ([raw file](https://github.com/cockroachdb/cockroach/blob/40d1991a7cda2a49f587c1b2432baf977b2763fd/pkg/util/container/ring/buffer_test.go#L202)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: isn't this a negative `minSize` -- what does that mean?
</blockquote></details>

See the other thread. It's a hack to disable hitting the lower bound, and make the stopping condition be "when `maxSize` is hit". It could have also been -1.

**Pull request merged**